java/ql/src/Security/CWE/CWE-918/RequestForgery.qll

```diff
@@ -0,0 +1,70 @@
+import java
+import semmle.code.java.dataflow.FlowSources
+import semmle.code.java.frameworks.javase.URI
+import semmle.code.java.frameworks.javase.URL
+import semmle.code.java.frameworks.javase.Http
+import semmle.code.java.frameworks.spring.Spring
+import semmle.code.java.dataflow.DataFlow
+
+module RequestForgery {
+  import RequestForgeryCustomizations::RequestForgery
+
+  /**
+   * A taint-tracking configuration for reasoning about request forgery.
+   */
+  class RequestForgeryRemoteConfiguration extends TaintTracking::Configuration {
+    RequestForgeryRemoteConfiguration() { this = "Server Side Request Forgery" }
+
+    override predicate isSource(DataFlow::Node source) { source instanceof RemoteFlowSource }
+
+    override predicate isSink(DataFlow::Node sink) { sink instanceof Sink }
+
+    override predicate isAdditionalTaintStep(DataFlow::Node pred, DataFlow::Node succ) {
+      additionalStep(pred, succ)
+    }
+  }
+
+  class RequestForgeryLocalConfiguration extends TaintTracking::Configuration {
+    RequestForgeryLocalConfiguration() { this = "Server Side Request Forgery" }
+
+    override predicate isSource(DataFlow::Node source) { source instanceof LocalUserInput }
+
+    override predicate isSink(DataFlow::Node sink) { sink instanceof Sink }
+
+    override predicate isAdditionalTaintStep(DataFlow::Node pred, DataFlow::Node succ) {
+      additionalStep(pred, succ)
+    }
+  }
+}
+
+predicate additionalStep(DataFlow::Node pred, DataFlow::Node succ) {
+  // propagate to a URI when its host is assigned to
+  exists(URIConstructor c | c.hostArg() = pred.asExpr() | succ.asExpr() = c)
+  or
+  // propagate to a URL when its host is assigned to
+  exists(URLConstructor c | c.hostArg() = pred.asExpr() | succ.asExpr() = c)
+  or
+  // propagate to a RequestEntity$BodyBuilder when its url is assigned to
+  exists(MethodAccess m |
+    m.getMethod().getDeclaringType() instanceof SpringRequestEntity and
+    (
+      m.getMethod().hasName(["get", "post", "head", "delete", "options", "patch", "put"]) and
+      m.getArgument(0) = pred.asExpr() and
+      m = succ.asExpr()
+    )
+    or
+    m.getMethod().hasName("method") and
+    m.getArgument(1) = pred.asExpr() and
+    m = succ.asExpr()
+  )
+  or
+  // propagate from a `RequestEntity<>$BodyBuilder` to a `RequestEntity`
+  // when the builder is tainted
+  exists(MethodAccess m, RefType t |
+    m.getMethod().getDeclaringType() = t and
+    t.hasQualifiedName("org.springframework.http", "RequestEntity<>$BodyBuilder") and
```

Comment:
Would be nice to turn this into a type instead of having it inline? Maybe?

Reply:
Not doing that. I don't see any reason why someone would want to reuse this. Any how, if this is required in the future, this can be refactored then.